### ISSUE TITLE
Add support for using external GPG instead of using built-in GPG

### ISSUE THREAD

**Issue by realYulin** [git-for-windows/git#3997]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.37.2.windows.2
cpu: x86_64
built from commit: c5597badf94101f0076e6667edee120ab1ee5bac
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.22622.586]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
> type "$env:USERPROFILE\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VisualStudioCode
Custom Editor Path: 
Default Branch Option:  
Path Option: Cmd
SSH Option: ExternalOpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Enabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

None

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Windows PowerShell

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
Suggestion for installer
```
 - What did you expect to occur after running these commands?

The installer should ask which GPG executable should be used. I have installed Gpg4win and use it generate a GPG key, but Git’s GPG is not compatible to Gpg4win’s GPG (Gpg4win stores GPG keys at C:\Users\USERNAME\AppData\Roaming\gnupg while Git stores it at  C:\Users\USERNAME\\.gnupg). Because I have enabled the "gpgsign" config, I have to export the private key and reimport it at Git Bash.

 - What actually happened instead?

Git use its own GPG so when I commit, I can always see this error "gpg: key "xxxxxxxxxxxxxxxx" not found: Skipped"

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

None

**Comment by dscho** [comment]
Sounds like a good idea. You will need to imitate git-for-windows/build-extra@68fdb43425e47cf956f80db0904e3733d0cb51b2 (which did the same for SSH). To do that:

1. install [Git for Windows' SDK](https://gitforwindows.org/#download-sdk),
2. `sdk cd installer`,
3. edit `install.iss`
4. build a new installer via `sdk build installer`/test the page via `./release.sh -d <new page>`
5. verify that that installer works as expected?
6. open a PR?

**Comment by Okeanos** [comment]
Personally, I would suggest using a workaround similar to what I do with OpenSSH as shown [here](https://github.com/Okeanos/dotfiles-windows/blob/main/stow/shell/dot-path#L2), i.e. ensure that the Git-Bash loads the desired SSH binaries before the supplied ones – I am fairly certain that it should work the same for gpg _if_ the executables have the same name between the different versions.

This has the added benefit that you can switch between them at will (which is most certainly a benefit for OpenSSH).

If time allows I may attempt to contribute a GPG patch for the installer, though (no promises).

**Comment by Okeanos** [comment]
I just created a [branch+commit](https://github.com/Okeanos/build-extra/commit/fed44d3e09172c553f82a9092571aca82876e5be) that enables "skipping" GPG during the Git installation. It builds and works as advertised, I think. I didn't have time to verify it in depth, though.

In particular I am not sure:

- whether all gpg related files are correctly skipped (I believe `gpg-error-something.exe` remains installed?!)
- whether all the gpg related files really should be skipped (there's way more than I initially thought).

I am not sure whether adding more pages to the Wizard in this fashion … is a good user experience. It'd probably be interesting to have something similar to the components page (where the Terminal fragment, Explorer integration, etc.) are presented and allow users to skip installation of these pieces (ssh, gpg, curl, …?) in a single dialog. I won't be implementing such a thing, though.

**Comment by Okeanos** [comment]
Yeah, I'll definitely open a PR (probably as WIP) soon. I just didn't get to it yesterday (mostly because of lack of testing on my part) but wanted to at least leave a note. It wasn't meant as a "please review"-message. More of a "I actually started working on it, feel free to have a look" sort of thing.

**Comment by filips123** [comment]
Sorry to bother, but is there a reason why this hasn't been implemented yet and that PR had been closed without being merged? My use-case for external GPG is that I want to use Gpg4win, which often provides newer versions and has better integration with some other tools.

In that PR, I see that there were some discussions about just setting `gpg.program` to the external GPG program and keeping the Git version bundled. However, I think that is not a good solution. Because then, two different GPG versions would exist on the system, so which is used would depend on the PATH order, which can be quite annoying, as sometimes the order changes during upgrades/installations, so it could cause problems if a different version is used. Besides that, not having two GPG installations would save some disk space.

**Comment by dscho** [comment]
> why this hasn't been implemented yet and that PR had been closed without being merged?

The explanation for that is in the discussion. It should be quite plain from the discussion that we couldn't come up with a robust solution, and therefore abandoned the effort.

**Comment by Okeanos** [comment]
> The explanation for that is in the discussion.

Indeed and from my side has not changed at all. I still think this is a worthwhile feature to have but the implementation and consequences are non-trivial. As a result I don't see myself being able to provide a satisfactory solution.
